### PR TITLE
Make ImportMeta.resolve conditional

### PR DESCRIPTION
**Pull request by sandersn**
The DOM type for ImportMeta.resolve is `(specifier: string): string`, and Typescript 5.6 will ship this type, which conflicts with the type in node/v16 and node/v18.

This PR makes the type conditional -- when the DOM is present, it will use the DOM type. When it's not, it will use the type as it is today.

I had to cut the tests down so they would pass whether or not the DOM was present.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/node/v16/module.d.ts
1. types/node/v18/module.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against e6cbf57deb54033ee2dc1547f4e21fd355f44cbf
</p>

File: types/node/v16/test/module.ts [hunk 1]
```diff
@@ -53,8 +53,8 @@ const entry: Module.SourceMapping = smap.findEntry(1, 1);
 {
     const importmeta: ImportMeta = {} as any; // Fake because we cannot really access the true `import.meta` with the current build target
     importmeta.url; // $ExpectType string
-    importmeta.resolve!("local", "/parent"); // $ExpectType Promise<string>
-    importmeta.resolve!("local", new URL("https://parent.module")); // $ExpectType Promise<string>
```
Comment (jakebailey) [inline types/node/v16/test/module.ts:57]: These tests can be moved to files that are only loaded by the non-DOM tsconfig, fwiw.

File: types/node/v16/module.d.ts [hunk 2]
```diff
@@ -195,8 +195,13 @@ declare module "module" {
         static Module: typeof Module;
         constructor(id: string, parent?: Module);
     }
+    type ImportMetaDOMCompat = typeof globalThis extends { onmessage: any } ? {
+    resolve(specifier: string): string;
```
Comment (jakebailey) [inline types/node/v16/module.d.ts:199]: I feel like this is going to be a hazard; if anyone includes DOM, they're going to be misled into thinking they don't need to await. But, if they're including DOM, maybe their "main" environment is DOM and they don't want to await?

I guess I don't know how to do any better.

Comment (andrewbranch) [reply types/node/v16/module.d.ts:199]: I think people with lib.dom.d.ts and @types/node are more likely to be running in the DOM than in Node.js. I.e., it’s easier to include @types/node unintentionally than it is to include the DOM lib unintentionally.

**Comment by jakebailey** [comment]
Hm, seems like the PR is missing the changes to the tsconfigs so these files get loaded?

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/node/v16/node-tests-dom.ts
1. types/node/v16/node-tests-non-dom.ts
1. types/node/v18/node-tests-dom.ts
1. types/node/v18/node-tests-non-dom.ts
1. types/node/v16/test/module-dom.ts

<details>
<summary>as well as these 3 other files...</summary>
<p>types/node/v16/test/module-non-dom.ts, types/node/v18/test/module-dom.ts, types/node/v18/test/module-non-dom.ts</p>
</details>


Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 2822e7bc2c886a2cc6d5ab7a584d62b20dce4fc0
</p>

File: types/node/v16/module.d.ts [hunk 3]
```diff
@@ -195,8 +195,14 @@ declare module "module" {
         static Module: typeof Module;
         constructor(id: string, parent?: Module);
     }
+    type ImportMetaDOMCompat = typeof globalThis extends { onmessage: any } ? {
+            resolve(specifier: string): string;
```
Comment (jakebailey) [inline types/node/v16/module.d.ts:199]: In the compat case, can you still define the form with the optional param returning a promise? Or maybe that's overkill.

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/node/v16/test/module-dom.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against f528dfc0012a68fb5ba71aa381c7b64fcef0624f
</p>

File: types/node/v16/module.d.ts (hunk 3 above)
Comment (sandersn) [reply types/node/v16/module.d.ts:199]: nope:

```
../../../../ts/built/local/lib.es5.d.ts:631:11 - error TS2430: Interface 'ImportMeta' incorrectly extends interface '{ resolve(specifier: string, parent?: string | URL | undefined): Promise<string>; }'.
  The types returned by 'resolve(...)' are incompatible between these types.
    Type 'string' is not assignable to type 'Promise<string>'.
```

**Review by jakebailey — APPROVED** [review]
It's unfortunate, but I think the best we can do to make things work.

**Pull request merged**